File: src/main/java/eu/alessiobianchi/websockettestserver/logic/RelayService.java
package eu.alessiobianchi.websockettestserver.logic;

import org.eclipse.jetty.websocket.api.Session;
import org.springframework.stereotype.Service;

import javax.annotation.PostConstruct;
import java.io.IOException;
import java.util.Set;
import java.util.concurrent.CopyOnWriteArraySet;

@Service("RelayService")
public class RelayService {
    private final Set<Session> listenerSessions = new CopyOnWriteArraySet<>();

    public void removeSession(Session session) {
        listenerSessions.remove(session);
    }

    public void addSession(Session session) {
        listenerSessions.add(session);
    }

    @PostConstruct
    private void keepAlive() {
        Thread thread = new Thread(() -> {
            while (true) {
                try {
                    Thread.sleep(30_000);
                    String message = "{}";
                    listenerSessions.stream()
                            .filter(Session::isOpen)
                            .forEach(session -> send(session, message));
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        });
        thread.start();
    }

    public void sendToAll(String message) {
        listenerSessions.stream()
                .filter(Session::isOpen)
                .forEach(session -> send(session, message));
    }

    private void send(Session session, String message) {
        try {
            session.getRemote().sendString(message);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/java/eu/alessiobianchi/websockettestserver/logic/RelayService.kt
package eu.alessiobianchi.websockettestserver.logic

import org.eclipse.jetty.websocket.api.Session
import org.springframework.stereotype.Service
import java.io.IOException
import java.util.concurrent.CopyOnWriteArraySet
import javax.annotation.PostConstruct

@Service("RelayService")
class RelayService {
    private val listenerSessions: MutableSet<Session> = CopyOnWriteArraySet()

    fun removeSession(session: Session) {
        listenerSessions.remove(session)
    }

    fun addSession(session: Session) {
        listenerSessions.add(session)
    }

    @PostConstruct
    private fun keepAlive() {
        val thread = Thread {
            while (true) {
                try {
                    Thread.sleep(30000)
                    val message = "{}"
                    listenerSessions.stream()
                        .filter { obj -> obj.isOpen }
                        .forEach { session -> send(session, message) }
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
            }
        }
        thread.start()
    }

    fun sendToAll(message: String) {
        listenerSessions.stream()
            .filter { obj -> obj.isOpen }
            .forEach { session -> send(session, message) }
    }

    private fun send(session: Session, message: String) {
        try {
            session.remote.sendString(message)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}